### PR TITLE
test(vm): Add tests for EVM emulator – multivm

### PR DESCRIPTION
**Pull request by slowli**
## What ❔

Adds unit tests for the `multivm` crate testing a mock EVM emulator.

## Why ❔

Ensures that EVM emulation will work as expected.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk_supervisor fmt` and `zk_supervisor lint`.

### CODE REVIEW

File: core/lib/multivm/src/versions/vm_latest/tests/evm_emulator.rs [hunk 1]
```diff
@@ -74,3 +92,229 @@ fn tracing_evm_contract_deployment() {
         evm_bytecode
     );
 }
+
+#[test]
+fn mock_emulator_basics() {
```
Comment (0xVolosnikov) [inline core/lib/multivm/src/versions/vm_latest/tests/evm_emulator.rs:97]: Why not use `build_vm` function here?

File: core/lib/multivm/src/versions/vm_latest/tests/evm_emulator.rs [hunk 2]
```diff
@@ -74,3 +92,229 @@ fn tracing_evm_contract_deployment() {
...
+    let mut vm = build_vm(deploy_emulator, recipient_address);
+    let account = &mut vm.rich_accounts[0];
+
+    let test_payment_fn = mock_emulator_abi.function("testPayment").unwrap();
+
+    let mut current_balance = U256::zero();
+    for i in 1_u64..=5 {
+        let transferred_value = (1_000_000_000 * i).into();
+        current_balance += transferred_value;
+        let tx = account.get_l2_tx_for_execute(
```
Comment (0xVolosnikov) [inline core/lib/multivm/src/versions/vm_latest/tests/evm_emulator.rs:190]: I'd propose to move this logic to the separate function

https://github.com/matter-labs/zksync-era/pull/3045/files#diff-36e5f0411039c083b35af6d6273d4dba7992a214a5e6330b45bc5a451d52a7b1R190-R205

File: core/lib/multivm/src/versions/vm_latest/tests/evm_emulator.rs [hunk 3]
```diff
@@ -74,3 +92,229 @@ fn tracing_evm_contract_deployment() {
...
+    let (_, vm_result) = vm
+        .vm
+        .execute_transaction_with_bytecode_compression(tx, true);
+    assert!(!vm_result.result.is_failed(), "{vm_result:?}");
+}
+
+#[test]
+fn mock_emulator_with_deployment() {
+    let mut storage = InMemoryStorage::with_system_contracts(hash_bytecode);
+    override_system_contracts(&mut storage);
```
Comment (0xVolosnikov) [inline core/lib/multivm/src/versions/vm_latest/tests/evm_emulator.rs:265]: `build_vm`?

File: etc/contracts-test-data/contracts/mock-evm/mock-evm.sol [hunk 4]
```diff
@@ -78,15 +78,119 @@ contract MockContractDeployer {
...
+}
+
+uint constant EVM_EMULATOR_STIPEND = 1 << 30;
+
+/**
+ * Mock EVM emulator used in low-level tests.
+ */
+contract MockEvmEmulator is IRecursiveContract {
+    IAccountCodeStorage constant ACCOUNT_CODE_STORAGE_CONTRACT = IAccountCodeStorage(address(0x8002));
+    address constant CODE_ORACLE_ADDR = address(0x8012);
```
Comment (0xVolosnikov) [inline etc/contracts-test-data/contracts/mock-evm/mock-evm.sol:111]: Unused?

File: etc/contracts-test-data/contracts/mock-evm/mock-evm.sol [hunk 5]
```diff
@@ -78,15 +78,119 @@ contract MockContractDeployer {
...
+    IRecursiveContract recursionTarget;
+
+    function recurse(uint _depth) public validEvmEntry returns (uint) {
+        require(gasleft() < 2 * EVM_EMULATOR_STIPEND, "stipend provided multiple times");
+
+        if (_depth <= 1) {
+            return 1;
+        } else {
+            IRecursiveContract target = (address(recursionTarget) == address(0)) ? this : recursionTarget;
+            // The real emulator deducts the stipend when performing far calls, so we emulate this behavior as well.
```
Comment (0xVolosnikov) [inline etc/contracts-test-data/contracts/mock-evm/mock-evm.sol:146]: It rather limits passed evm gas, not deducts stipend

**Comment by 0xVolosnikov** [comment]
It would be also nice to test delegatecall, staticcall, "cross-VM" calls (EVM -> non-EVM -> EVM, etc)

**Comment by slowli** [comment]
> "cross-VM" calls (EVM -> non-EVM -> EVM, etc)

They are tested in this way in `testDeploymentAndCall`. Will try to test non-EVM -> EVM -> non-EVM calls similarly.

**Review by 0xVolosnikov — APPROVED** [review]

**Pull request merged**